src/allocator.rs

```diff
@@ -332,13 +336,36 @@ impl Allocator {
 
     pub fn new_pair(&mut self, first: NodePtr, rest: NodePtr) -> Result<NodePtr, EvalErr> {
         let idx = self.pair_vec.len();
-        if idx == MAX_NUM_PAIRS {
+        if idx >= self.max_num_pairs {
             return err(self.nil(), "too many pairs");
         }
         self.pair_vec.push(IntPair { first, rest });
         Ok(NodePtr::new(ObjectType::Pair, idx))
     }
 
+    pub fn reduce_pair_max(&mut self, amount: usize) -> Result<(), EvalErr> {
+        if self.max_num_pairs - amount <= self.pair_vec.len() {
```

Comment:
I think this risks wrapping, doesn't it? How about this instead:
```suggestion
        if self.max_num_pairs - self.pair_vec.len() < amount {
```

Reply:
if `max_num_pairs == pair_vec.len()`, that's fine. No failure. See comment above

Reply:
Done